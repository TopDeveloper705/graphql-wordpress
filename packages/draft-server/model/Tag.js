import DataLoader from 'dataloader';
import findByIds from 'mongo-find-by-ids';
import slugify from '../utils/slugify';

export default class Tag {
  constructor(context) {
    this.context = context;
    this.collection = context.db.collection('tag');
    this.loader = new DataLoader(ids => findByIds(this.collection, ids));
  }

  findOneById(id) {
    return this.loader.load(id);
  }

  count(args = {}) {
    const criteria = Object.assign({}, args);
    delete criteria.search;
    if (args.search) {
      criteria.$text = { $search: args.search };
    }
    return this.collection.find(criteria).count();
  }

  all({ limit = 10, offset = 0, search = null }) {
    const criteria = {};
    if (search) {
      criteria.$text = { $search: search };
    }

    return this.collection
      .find(criteria)
      .sort({ name: 1 })
      .skip(offset)
      .limit(limit)
      .toArray();
  }

  async insert(doc) {
    const slug = slugify(doc.name);
    const docToInsert = Object.assign({}, doc, {
      _id: slug,
      slug,
      createdAt: Date.now(),
      updatedAt: Date.now(),
    });
    const id = (await this.collection.insertOne(docToInsert)).insertedId;
    return id;
  }

  async updateById(id, doc) {
    const docToUpdate = Object.assign({}, doc);
    const ret = await this.collection.update(
      { _id: id },
      {
        $set: Object.assign({}, docToUpdate, {
          updatedAt: Date.now(),
        }),
      }
    );
    this.loader.clear(id);
    return ret;
  }

  async removeById(id) {
    const ret = this.collection.remove({ _id: id });
    this.loader.clear(id);
    return ret;
  }
}
