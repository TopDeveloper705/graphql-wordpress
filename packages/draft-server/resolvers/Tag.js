import { parseConnection } from '../utils/collection';

const resolvers = {
  Tag: {
    id(tag) {
      return tag._id;
    },
  },
  Query: {
    async tags(root, args, { Tag }) {
      return parseConnection(Tag, args);
    },

    tag(root, { id }, { Tag }) {
      return Tag.findOneById(id);
    },
  },
  Mutation: {
    async createTag(root, { input }, { Tag }) {
      const id = await Tag.insert(input);
      return Tag.findOneById(id);
    },

    async updateTag(root, { id, input }, { Tag }) {
      await Tag.updateById(id, input);
      return Tag.findOneById(id);
    },

    removeTag(root, { id }, { Tag }) {
      return Tag.removeById(id);
    },
  },
};

export default resolvers;
