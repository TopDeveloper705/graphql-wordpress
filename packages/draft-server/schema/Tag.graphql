enum TAXONOMY {
  ARTIST
  EVENT
  LOCATION
  PROGRAM
}

type Tag {
  id: String
  name: String
  slug: String
  taxonomy: [TAXONOMY]
}

type TagEdge {
  node: Tag
  cursor: String
}

type TagConnection {
  count: Int
  edges: [TagEdge]
  pageInfo: PageInfo
}

extend type Query {
  tags(
    first: Int
    after: String
    last: Int
    before: String
    taxonomy: String
    search: String
  ): TagConnection
  tag(id: String): Tag
}

input CreateTagInput {
  name: String
  taxonomy: [TAXONOMY]
}

input UpdateTagInput {
  name: String
  taxonomy: [TAXONOMY]
}

extend type Mutation {
  createTag(input: CreateTagInput!): Tag
  updateTag(id: String!, input: UpdateTagInput!): Tag
  removeTag(id: String!): Boolean
}
