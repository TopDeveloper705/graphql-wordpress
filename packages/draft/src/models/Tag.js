import Model from './Model';
import { getUniqueSlug } from './utils';

export default class Tag extends Model {
  constructor(context) {
    super(context);

    this.collection = context.db.collection('tag');
  }

  all({ limit = 10, offset = 0, taxonomy = null, search = null }) {
    const criteria = {};
    if (taxonomy) {
      criteria.taxonomy = taxonomy;
    }
    if (search) {
      criteria.$text = { $search: search };
    }

    return this.collection
      .find(criteria)
      .sort({ name: 1 })
      .skip(offset)
      .limit(limit)
      .toArray();
  }

  async insert(doc) {
    const slug = await getUniqueSlug(this.collection, doc.name);
    const docToInsert = Object.assign({}, doc, {
      _id: slug,
      slug,
      taxonomy: [],
      createdAt: Date.now(),
      updatedAt: Date.now(),
    });
    const id = (await this.collection.insertOne(docToInsert)).insertedId;
    return id;
  }
}
