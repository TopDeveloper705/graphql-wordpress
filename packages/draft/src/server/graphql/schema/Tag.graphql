enum TAXONOMY {
  ARTIST
  EVENT
  LOCATION
  PROGRAM
}

type Tag {
  id: String
  name: String
  slug: String
  taxonomy: [TAXONOMY]
}

type TagEdge {
  node: Tag
  cursor: String
}

type TagConnection {
  count: Int
  edges: [TagEdge]
  pageInfo: PageInfo
}

input CreateTagInput {
  name: String
  taxonomy: [TAXONOMY]
}

input UpdateTagInput {
  name: String
  taxonomy: [TAXONOMY]
}

extend type Query {
  tags(
    first: Int
    after: String
    last: Int
    before: String
    taxonomy: String
    search: String
  ): TagConnection
  tag(id: String): Tag
}

extend type Mutation {
  createTag(input: CreateTagInput!): Tag
  updateTag(id: String!, input: UpdateTagInput!): Tag
  removeTag(id: String!): Boolean
}
