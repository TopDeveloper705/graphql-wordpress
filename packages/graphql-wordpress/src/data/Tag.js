import { toGlobalId } from 'graphql-relay';

class Tag {
  getID() {
    return toGlobalId(this.constructor.name, this.id);
  }

  static getEndpoint() {
    return process.env.WP_TAGS_ENDPOINT || 'wp/v2/tags';
  }
}

export default Tag;
