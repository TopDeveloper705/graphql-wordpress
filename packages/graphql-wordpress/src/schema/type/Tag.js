import { GraphQLObjectType } from 'graphql';
import TermInterface from 'interface/Term';
import { globalIdField, slug, name, link } from 'field/identifier';
import taxonomy from 'field/taxonomy';
import description from 'field/description';
import metaField from 'field/meta';
import { registerNodeType, NodeInterface } from 'type/relayNode';

const TagType = new GraphQLObjectType({
  name: 'Tag',
  description: 'A unique identifier for a post.',
  interfaces: [TermInterface, NodeInterface],
  isTypeOf(term) {
    return term.taxonomy === 'post_tag';
  },
  fields: {
    id: globalIdField(),
    ...description,
    ...link,
    ...name,
    ...slug,
    ...taxonomy,
    meta: metaField(),
  },
});

registerNodeType(TagType);

export default TagType;
