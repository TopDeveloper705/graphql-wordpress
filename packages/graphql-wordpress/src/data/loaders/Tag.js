import Dataloader from 'dataloader';
import fetchData from 'data/utils';
import Tag from 'data/Tag';

// Dataloader expects IDs that can be read by the REST API

export default function getTagLoaders() {
  const endpoint = Tag.getEndpoint();

  const tagLoader = new Dataloader(ids =>
    fetchData(endpoint, {
      qs: { include: ids, per_page: 100, orderby: 'include' },
    }).then(({ data: { body } }) => body)
  );
  const slugLoader = new Dataloader(slugs =>
    fetchData(endpoint, { qs: { slug: slugs, per_page: 100 } })
      .then(({ data: { body } }) => body)
      // the REST API does not order by FIELD(slug, ....) yet
      .then(tags => slugs.map(slug => tags.find(tag => slug === tag.slug)))
  );

  return {
    load: async id => {
      const data = await tagLoader.load(id);
      return data ? Object.assign(new Tag(), data) : null;
    },

    loadMany: async ids => {
      const data = await tagLoader.loadMany(ids);
      return data ? data.map(entry => Object.assign(new Tag(), entry)) : null;
    },

    loadBySlug: async slug => {
      const data = await slugLoader.load(slug);
      return data ? Object.assign(new Tag(), data) : null;
    },
  };
}
