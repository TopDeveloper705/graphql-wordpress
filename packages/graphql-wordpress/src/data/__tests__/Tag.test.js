import Tag from 'data/Tag';

const backupEnv = Object.assign({}, process.env);

afterAll(() => {
  process.env = backupEnv;
});

describe('Test Tag data access', () => {
  test('Get endpoint', () => {
    expect(Tag.getEndpoint()).toMatchSnapshot();
  });

  test('Get backup endpoint', () => {
    delete process.env.WP_TAGS_ENDPOINT;
    expect(Tag.getEndpoint()).toMatchSnapshot();
  });
});
