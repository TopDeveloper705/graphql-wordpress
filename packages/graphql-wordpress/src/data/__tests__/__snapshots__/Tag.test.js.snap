// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Tag data access Get backup endpoint 1`] = `"wp/v2/tags"`;

exports[`Test Tag data access Get endpoint 1`] = `"wp/v2/tags"`;
